core/src/test/scala/org/apache/spark/sql/delta/DeltaCDCSuite.scala

```diff
@@ -0,0 +1,631 @@
+
+  test("table schema changed after dataframe with ending not specified") {
+    withTempDir { dir =>
+      val path = dir.getAbsolutePath
+      spark.range(5).selectExpr("id", "'text' as text")
+        .write.format("delta").save(path)
+      val cdcResult = cdcRead(new TablePath(path), StartingVersion("0"), Unbounded)
+      sql(s"ALTER TABLE delta.`$path` ADD COLUMN (newCol STRING)")
+      sql(s"INSERT INTO delta.`$path` VALUES (5, 'text', 'newColVal')")
+
+      // Just ignoring the new column is pretty weird, but it's what we do for non-CDC dataframes,
+      // so we preserve the behavior rather than adding a special case.
+      checkAnswer(
+        cdcResult.selectExpr("id", "_change_type", "_commit_version"),
+        Row(0, "insert", 0) :: Row(1, "insert", 0) :: Row(2, "insert", 0) ::
+          Row(3, "insert", 0) :: Row(4, "insert", 0) :: Row(5, "insert", 2) :: Nil
+      )
+    }
+  }
+
+  test("An error should be thrown when CDC is not enabled") {
+    val tblName = "tbl"
+    withTable(tblName) {
+      withSQLConf(DeltaConfigs.CHANGE_DATA_FEED.defaultTablePropertyKey -> "false") {
+        // create version with cdc disabled - v0
+        spark.range(10).write.format("delta").saveAsTable(tblName)
+      }
+      val deltaTable = io.delta.tables.DeltaTable.forName(tblName)
+      // v1
+      deltaTable.delete("id > 8")
+
+      // v2
+      sql(s"ALTER TABLE ${tblName} SET TBLPROPERTIES " +
+        s"(${DeltaConfigs.CHANGE_DATA_FEED.key}=true)")
+
+      // v3
+      spark.range(10, 20).write.format("delta").mode("append").saveAsTable(tblName)
+
+      // v4
+      deltaTable.delete("id > 18")
+
+      // v5
+      sql(s"ALTER TABLE ${tblName} SET TBLPROPERTIES " +
+        s"(${DeltaConfigs.CHANGE_DATA_FEED.key}=false)")
+
+      var e = intercept[AnalysisException] {
+        cdcRead(new TableName(tblName), StartingVersion("0"), EndingVersion("4")).collect()
+      }
+      assert(e.getMessage === DeltaErrors.changeDataNotRecordedException(0, 0, 4).getMessage)
+
+      val cdcDf = cdcRead(new TableName(tblName), StartingVersion("2"), EndingVersion("4"))
+      assert(cdcDf.count() == 11) // 10 rows inserted, 1 row deleted
+
+      // Check that we correctly detect CDC is disabled and fail the query for multiple types of
+      // ranges:
+      //  * disabled at the end but not start - (2, 5)
+      //  * disabled at the start but not end - (1, 4)
+      //  * disabled at both start and end (even though enabled in the middle) - (1, 5)
+      for ((start, end, firstDisabledVersion) <- Seq((2, 5, 5), (1, 4, 1), (1, 5, 1))) {
+        e = intercept[AnalysisException] {
+          cdcRead(
+            new TableName(tblName),
+            StartingVersion(start.toString), EndingVersion(end.toString)).collect()
+        }
+        assert(e.getMessage === DeltaErrors.changeDataNotRecordedException(
+          firstDisabledVersion, start, end).getMessage)
+      }
+    }
+  }
+}
+
+class DeltaCDCSuite extends DeltaCDCSuiteBase {
```

Comment:
why this structure? why are some tests in the base and some here?

Reply:
I will rename this suite. this is the Scala test suite. in the future, we may support SQL syntax and will want a SQL child test suite.